# Guided Project: Exploring Ebay Car Sales Data


In this project I'll work with a dataset of used car listings from the German Ebay's 'classifieds' section. The original dataset was uploaded to Kaggle, but has been  sampled and dirtied by Dataquest for this project. The aim is to clean and analyze the dataset, while getting familiair with using pandas in Jupyter notebook. So...uhm, let's go!


In [1]:
# importing libraries
import numpy as np
import pandas as pd

In [2]:
# reading csv
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
# exploring dataset
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## First observations

The dataset we'll be working with has 50,000 rows, each representing a single used car. The index axis has integer labels and each row contains a timestamp for when the information was added to the dataset, but the data doesn't appear to be sorted in any particular order. It contains information on the seller, price, make and model of the car, damage, date the ad was posted and the last time it was seen.

Looking a bit more closely, it seems I have my work cut out for me cleaning the data. The column labels, dates and timestamps all need cleaning and reformatting and some of the columns appear to contain only one value and may be dropped. Also, five of the columns contain null values and a few columns, like price can be converted to either floats or integers.

In [5]:
# column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# clean column names
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [7]:
# checking the result
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### What I just did and why

I've changed the column labels from camelcase to snakecase, so I won't have to worry about capitalization later on when calling specific columns. Besides that I've changed a few of the labels to make them easier to read and remember.

In [8]:
# some more exploration
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

## To Do List

1. Candidate columns to drop (contain mostly just 1 value):
    - seller
    - offer_type
    - nr_of_pictures
2. Columns to investigate:
    - price - 1421 listings have a price of $0
    - registration_year - The min and max value in this column are 1000 and 9999
    - power_ps - it's values range from 0 to 17700
3. Columns containing numeric information stored as text:
    - price
    - odometer

In [9]:
# converting the 'price' and 'odometer' columns to numeric values
autos["price"] = (autos["price"]
                 .str.strip()
                 .str.replace("$", "")
                 .str.replace(",", "")
                 .astype(int)
                 )

autos["odometer"] = (autos["odometer"]
                    .str.strip()
                    .str.replace("km", "")
                    .str.replace(",", "")
                    .astype(int)
                    )

autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [10]:
# exploring the price column
autos["price"].unique().shape

(2357,)

In [11]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [12]:
autos["price"].value_counts().head(15)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64

In [13]:
autos["price"].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

In [14]:
# exploring the odometer_km column
autos["odometer_km"].unique().shape

(13,)

In [15]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [16]:
autos["odometer_km"].value_counts().head()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

In [17]:
autos["odometer_km"].value_counts().sort_index(ascending=True).head()

5000     967
10000    264
20000    784
30000    789
40000    819
Name: odometer_km, dtype: int64

### Outliers

Looking at the price and odometer_km columns, only the price column appears to show outliers. There are quite a few listings with a price of 0 dollars or over 100,000 dollars. Unfortunately, the instructions for this project only say to remove the outliers and the rows containing the outliers, instead of replacing the values with None thus keeping the data in the other columns. So before I start cutting away I'd like to take a closer look at these outliers.

In [18]:
# selecting only the rows with a price of 0
autos_zero_price = autos.loc[autos["price"] == 0]

In [19]:
autos_zero_price

date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
502    2016-03-25 23:55:24                               C32_AMG_C55_C43_w203   
519    2016-03-28 08:53:58  Verkaufe/Tausche_Seat_Inca__VW_Caddy__aus_2002...   
565    2016-03-20 20:49:36     Tausche_Audi_S3_8p_gegen_Golf_VI_Gti_mit_xenon   
593    2016-04-01 21:44:15                              CLK_200__BRABUS_Optik   
660    2016-03-29 21:44:40                                       Opel_Corsa_B   
661    2016-03-18 22:49:48               Renault_Megane_Scenic_1.6_Tuev_12/17   
...                    ...                                                ...   
48882  2016-03-14 23:58:47                                         SUCHE_AUTO   
48889  2016-04-01 10:50:52                           Audi_A4_____TueV_09.2016   
48891  2016-03-20 22:50:14        MB_E_220_CDI__TAUSCHEN_nur_SUV/GELÄNDEWAGEN   
48982  2016-03-08 12:48:33                           Bmw_525D_Sport_Pro..Navi   
48985  2016-03-27 08:55:25                             Stockcar_Autos_Peugeot   
49043  2016-03-07 11:48:55                              Opel_Vectra_B_Caravan   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e

In [20]:
# these seem like legit listings, now for those of over 100,000
autos_high_price = autos.loc[autos["price"] > 100000]
autos_high_price

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
1878   2016-03-12 16:58:37                                  Porsche_911_Turbo   
2454   2016-03-21 22:51:29                                    Porsche_911_GT3   
2751   2016-03-15 10:52:35                               Porsche_911___993_4S   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7402   2016-03-22 19:48:09  Porsche_911_Carrera_4S_Cabrio_PDK__BOSE__NEU__...   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
8232   2016-04-01 21:50:47               Porsche_993_S_Schalter_BRD_neuwertig   
10500  2016-03-17 12:56:38                                        Porsche_991   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
11433  2016-03-30 21:44:21  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
14268  2016-04-01 12:52:27   Porsche_997_GT3_Approved_Lift_Unfallfrei_Mj.2011   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
16964  2016-04-01 16:46:18                    Bentley_Continental_Supersports   
17140  2016-03-05 17:39:22  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
17540  2016-03-30 00:49:49   Porsche_930_Turbo_classic_analytics_Gutachten_2+   
18509  2016-03-27 18:43:02                                    Porsche_991_GT3   
20351  2016-03-22 17:47:59                                    BMW_Z8_roadster   
21783  2016-03-26 22:06:24               Porsche_911_Carrera_4S_Cabriolet_PDK   
22060  2016-03-09 00:58:24  Tesla_Model_X90D_Autopilot_Leder_AHK_Kaltwette...   
22673  2016-03-05 18:40:13  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Club...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
28090  2016-03-08 19:50:06  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
29286  2016-03-22 15:58:56  Porsche_911_Turbo_PDK__Sport_Chrono_Garantie_D...   
32185  2016-03-25 12:42:09  Porsche_997_GT3_MKII_EIN_SAMMLERSTÜCK_IN_VOLLA...   
32840  2016-03-22 08:55:27  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
33638  2016-03-24 12:54:19                                 Porsche_911_GT3_RS   
33884  2016-03-26 21:55:12               Porsche_911_Carrera_4S_Cabriolet_PDK   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
36818  2016-03-27 18:37:37                                        Porsche_991   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
37840  2016-03-21 10:50:12                                        Porsche_997   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
38814  2016-03-22 16:53:44  Porsche_Porsche_964_TURBO_S_X33_WLS_ab_Werk_35...   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
40918  2016-03-20 18:40:05                             Porsche_911_991_GT3_RS   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
43282  2016-04-03 18:42:28               Porsche_911_Carrera_4S_Cabriolet_PDK   
43668  2016-03-16 18:47:26  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
44406  2016-03-15 14:47:26  Mercedes_Benz_170_S_Cabriolet_A__Startklar_fue...   
45387  2016-03-18 16:41:22  Porsche_911_GT3__Clubsp._Lift_nur_bis_25.03._v...   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6

### Cutting away

After looking more closely at the columns containing outliers in the price column, I can see that most of the listings are valid. I've decided only to exclude the rows with a price of over 350,000. With the exception of a few Ferrari's, these listings have strange values in the price column. 

In [21]:
autos = autos[autos["price"] < 350000]

### Now let's take a look at those timestamps

There are three columns containing timestamps: date_crawled, ad_created and last_seen. Two of these, date_crawled and last_seen, have been added by the crawler that created it and ad_created was crawled from the website. The last_seen column is the last time the ad was seen by the crawler. All timestamps are currently represented as strings and we'll have to convert it into a numerical representation. The registration_year and registration_month columns also represent date values but only contain integers, so we can look at them more easily.

In [22]:
date_crawled = (autos["date_crawled"]
                .str.strip()
                .str[:10]
                .value_counts(normalize=True, dropna=False)
                .sort_index()
               )

ad_created = (autos["ad_created"]
              .str.strip()
              .str[:10]
              .value_counts(normalize=True, dropna=False)
              .sort_index()
             )

last_seen = (autos["last_seen"]
             .str.strip()
             .str[:10]
             .value_counts(normalize=True, dropna=False)
             .sort_index()
            )

date_crawled

2016-03-05    0.025388
2016-03-06    0.013944
2016-03-07    0.035971
2016-03-08    0.033270
2016-03-09    0.033210
2016-03-10    0.032130
2016-03-11    0.032490
2016-03-12    0.036771
2016-03-13    0.015565
2016-03-14    0.036631
2016-03-15    0.033990
2016-03-16    0.029509
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037491
2016-03-22    0.032910
2016-03-23    0.032390
2016-03-24    0.029109
2016-03-25    0.031750
2016-03-26    0.032490
2016-03-27    0.031029
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033630
2016-03-31    0.031910
2016-04-01    0.033810
2016-04-02    0.035411
2016-04-03    0.038692
2016-04-04    0.036491
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

The distribution of dates in this column seems fairly even at first glance but there are a few outliers, most notably for the last three dates.

In [23]:
ad_created

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033230
2016-03-10    0.031870
2016-03-11    0.032790
2016-03-12    0.036611
2016-03-13    0.016925
2016-03-14    0.035231
2016-03-15    0.033750
2016-03-16    0.030009
2016-03-17    0.031189
2016-03-18    0.013724
2016-03-19    0.033850
2016-03-20    0.037871
2016-03-21 

In [24]:
last_seen

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015885
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015745
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018586
2016-03-24    0.019566
2016-03-25    0.019206
2016-03-26    0.016965
2016-03-27    0.016005
2016-03-28    0.020846
2016-03-29    0.022327
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023107
2016-04-02    0.024887
2016-04-03    0.025368
2016-04-04    0.024627
2016-04-05    0.124277
2016-04-06    0.220986
2016-04-07    0.130959
Name: last_seen, dtype: float64

In [25]:
autos["registration_year"].describe()

count    49985.000000
mean      2005.075503
std        105.728207
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The minimum and maximum value in registration_year shows there are incorrect values in this column which will need to be removed.

In [26]:
autos["registration_year"].value_counts().sort_index()

1000       1
1001       1
1111       1
1500       1
1800       2
1910       9
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       5
1957       2
1958       4
1959       7
1960      33
1961       6
1962       4
1963       9
        ... 
2001    2702
2002    2533
2003    2727
2004    2737
2005    3015
2006    2707
2007    2304
2008    2231
2009    2097
2010    1597
2011    1634
2012    1323
2013     806
2014     665
2015     399
2016    1315
2017    1452
2018     491
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: registration_year, Length: 97, dtype: int64

Since the dataset was scraped in 2016, any value higher than that should be inaccurate but because there seem to be quite a few listings with 2017 or 2018, I want to take a closer look at these before I remove them. With the Ford Model T going in production in 1908, I think it's safe to remove any listing with a registration_year before 1910 from this set.

In [27]:
reg_year_2017 = autos.loc[autos["registration_year"] == 2017]
reg_year_2017

date_crawled                                               name  \
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
65     2016-04-04 19:30:39                      Ford_Fiesta_zum_ausschlachten   
68     2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
113    2016-04-03 14:58:29                              Golf_4_Anfaenger_auto   
197    2016-04-05 10:36:24                              VW_Polo_9N_an_Bastler   
253    2016-03-27 13:25:18             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017   
348    2016-03-17 20:58:24  VW_Beetle_1.8Turbo_mit_Vollausstattung_und_seh...   
438    2016-03-10 20:36:25                VW_Golf_5_V_1.6_2004_Klima_Tempomat   
454    2016-03-25 09:57:03    Ford_Fiesta_Servo_D3_Kat_86€_Steuern_65.000_tkm   
457    2016-03-15 17:48:39  Mercedes_Benz_W203_C200_Kombi_Kompressor_c_kla...   
477    2016-03-31 08:53:34                                         Citroën_C4   
545    2016-03-22 18:50:53                                      Bmw_x5_/__E53   
548    2016-03-10 17:44:57            Peugeot_206_1.9_Hdi_4_Tueren_Tuev_02.17   
557    2016-03-16 03:36:34                           Smart_Fortwo_Passion_MHD   
568    2016-03-18 09:48:44             PRIMA_ANFÄNGERAUTO_smart_fortwo_cabrio   
577    2016-04-03 14:49:09                                  Volkswagen_Golf_3   
582    2016-03-31 11:57:46                                    V10_TDI_TOUAREG   
654    2016-04-01 15:48:34              Ich_verkaufe_meinen_gepflegten_daewoo   
740    2016-04-02 19:49:46                            Seat_Leon_zu_verkaufen!   
754    2016-03-26 13:47:56                         Ford_Ka_Kleinwagen_BJ_2008   
758    2016-03-12 18:42:33                            Honda_Civic_mit_TÜV_!!!   
765    2016-03-26 11:52:53               BMW_3er_Limousine_mit_1_Jahr_TÜV!!!!   
871    2016-04-04 19:38:38                                   Citroen_C2_60_PS   
889    2016-03-12 16:52:19                          Verkaufe_Opel_Corsa_C_1_2   
910    2016-03-15 17:52:50                         Renault_Kangoo_bj.2007_LPG   
1030   2016-03-24 20:47:08                     Ford_Galaxy___defekt_bastler__   
1040   2016-03-30 19:53:16                                         Audi_80_B4   
1066   2016-03-06 15:49:03               FORD_ESCORT_KOMBI/TURNIER_TÜV_7.2017   
1084   2016-03-13 16:42:45                                 Golf_3_Cabrio_1.8l   
...                    ...                                                ...   
48975  2016-03-21 13:44:47                                 Renault_Espace_iii   
48980  2016-03-28 19:50:38                                     Toyota_Yaris!!   
49003  2016-03-12 12:52:08                                 Renault_1.Hand_2.0   
49023  2016-03-23 00:49:23                                       Auto_Audi_A3   
49053  2016-03-15 15:57:37                     VOLKSWAGEN_GOLF_1.9_TDI_110_PS   
49054  2016-03-11 23:37:19                            Skoda_Fabia_aus_1._Hand   
49076  2016-03-25 21:39:21                             Omega_B_2.0_16v_136_Ps   
49108  2016-03-14 22:37:20                  TDI_20_L_DIESEL140_PS..AUTOMATICK   
49130  2016-03-09 18:46:05                                      Renault_Modus   
49154  2016-03-24 17:51:22                    Auto_mit_TÜV_bis_September_2017   
49245  2016-03-26 08:32:30                         Audi_A3_1.8T_tausch_moegl.   
49259  2016-03-19 18:45:05                  Volkswagen_Lupo_TÜV_06.2017_Klima   
49261  2016-03-10 12:51:40                                        Toyota_aygo   
49266  2016-03-10 00:54:09  Toyota_Corolla_Verso_7_Sitzer_104.000km_2_2_Di...   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49343  2016-03-12 11:56:55                                   bmw_320i_touring   
49347  2016-03-21 20:41:09                         Honda_civic_mit_TÜV_3/2017   
49389  2016-03-10 07:54:10                   

In [28]:
reg_year_2018 = autos.loc[autos["registration_year"] == 2018]
reg_year_2018

date_crawled                                               name  \
84     2016-03-27 19:52:54                                     Renault_twingo   
164    2016-03-13 20:39:16  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...   
390    2016-03-25 12:59:06   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19   
802    2016-03-19 11:37:23                 Lada_mit_wenig_km_neuem_Tuev_bj_08   
918    2016-03-15 15:53:56                              BMW_114i_Sondermodell   
1121   2016-03-26 21:38:21  Opel_Zafira_A_1.8_Z18XE_HU_02/18_164Tkm_Alu_Klima   
1551   2016-03-08 07:54:36            Verkaufe_hier_meine_Auto_vw_Passat_2007   
1889   2016-03-11 14:36:58                                        Golf_2_90PS   
1903   2016-04-05 06:56:26                               Tausche_Ford_Granada   
2006   2016-03-15 16:46:49                        Renault_Twingo_1.2_TÜV_neue   
2051   2016-03-21 13:40:00                                     Wartburg_353_W   
2089   2016-03-10 11:38:21                 Audi_A4_2_5_v6_NEU_TÜV_&Angemeldet   
2093   2016-03-12 14:51:41                        Sehr_gepflegter_Fiesta_2007   
2545   2016-03-27 15:37:02                                      VW_Golf5__GTI   
2642   2016-03-07 16:38:32                   Seat_arosa_HU_neu!!!_Top_Zustand   
2704   2016-03-31 11:50:17  Audi_A6_1_8_T_mit_LPG_TÜV_3/18_Neu_mit_Scheckh...   
2785   2016-03-30 12:55:06                                          Seat_Leon   
2886   2016-03-09 01:00:45                       Gute_Zustand__TÜV___03/_2018   
2960   2016-03-15 19:56:49               FIAT_STILO_1.6_16v_TÜV_03.2018_Klima   
2963   2016-03-19 20:51:45                              Audi_80s_zu_verkaufen   
3033   2016-03-22 14:51:12       VW_FOX_1_2L_Benzin_Scheckheft_gepflegt_2999€   
3231   2016-04-02 21:45:26             Golf_3_Cabrio_Blickfang_Scirocco_gruen   
3235   2016-03-28 20:38:11                            Bmw730_d___zu_verkaufen   
3392   2016-03-22 21:58:48                     Verkaufe_gepflegten_Fiat_Panda   
3399   2016-03-21 19:58:32                               Verkaufe_VW_Polo_6N2   
3591   2016-03-30 20:42:34         VW_Passat_1_6_FSI_TÜV_neu_in_super_Zustand   
3716   2016-04-01 19:47:45              Nissan_Micra_1.2_Tuev_Neu_Top_Zustand   
3801   2016-03-16 11:37:41  VERKAUFE_Nissan_Xtrail_2.2_Diesel_Gruene_plake...   
3848   2016-03-15 22:36:42                              BMW_316i___Frisch_TÜV   
3980   2016-03-24 17:57:43                Citroen_C5__DIESEL__Gruen___TÜV_NEU   
...                    ...                                                ...   
46530  2016-03-05 14:13:18                                      Sehr_gut_opel   
46716  2016-03-31 17:57:07                                          Auto_Soda   
46809  2016-03-25 21:40:17                                         Opel_corsa   
46811  2016-03-08 15:59:20                          Golf_4_1.6_Tuv_bis_1.2018   
46986  2016-03-22 22:55:30                            Skoda_Octavia_Combi_Tsi   
47016  2016-03-23 17:44:53                             Seat_Leon_1M__1_6L_16v   
47183  2016-03-19 22:36:59                                     Renault_Laguna   
47258  2016-03-18 21:43:25                                        Jaguar_xj40   
47305  2016-03-10 15:48:33                               Renault_Clio_1.2_TOP   
47312  2016-03-16 16:44:07       Nur_heute_4000€_Audi_A4__TÜV_NEU_3/18_mit_18   
47340  2016-03-13 05:36:20  Golf_4_Cabrio_1.6_mit_Automatikgetriebe_und_Tu...   
47367  2016-03-17 19:55:53                                   BMW_318i_compact   
47521  2016-03-12 10:51:25   VW_Caddy_Modell_2008_mit_sehr_guter_Ausstattung.   
47596  2016-03-08 15:37:02                                     Volkswagen_fox   
47757  2016-04-03 15:06:18                          Ford_Fiesta_1_6_;_101_PS!   
47809  2016-04-01 19:47:47                               Opel_corsa_tuev_2018   
47823  2016-04-02 23:51:51               Ford_Focus_mk1_1_8_l_TDDi_HU_02/2018   
47876  2016-03-09 17:51:02                   

As you'll notice, in both reg_year_2017 and reg_year_2018 all vehicle_type values appear to be NaN. Let's see.

In [29]:
reg_year_2017.describe(include="all")

date_crawled        name  seller offer_type         price  \
count                  1452        1452    1452       1452   1452.000000   
unique                 1450        1411       1          1           NaN   
top     2016-03-21 07:56:38  Opel_Astra  privat    Angebot           NaN   
freq                      2           4    1452       1452           NaN   
mean                    NaN         NaN     NaN        NaN   3402.242424   
std                     NaN         NaN     NaN        NaN   4330.582019   
min                     NaN         NaN     NaN        NaN      0.000000   
25%                     NaN         NaN     NaN        NaN    899.000000   
50%                     NaN         NaN     NaN        NaN   1850.000000   
75%                     NaN         NaN     NaN        NaN   4499.000000   
max                     NaN         NaN     NaN        NaN  48500.000000   

       abtest vehicle_type  registration_year  gearbox     power_ps model  \
count    1452            1             1452.0     1272  1452.000000  1248   
unique      2            1                NaN        2          NaN   167   
top      test        kombi                NaN  manuell          NaN  golf   
freq      751            1                NaN     1060          NaN   165   
mean      NaN          NaN             2017.0      NaN    92.442149   NaN   
std       NaN          NaN                0.0      NaN   258.813079   NaN   
min       NaN          NaN             2017.0      NaN     0.000000   NaN   
25%       NaN          NaN             2017.0      NaN     0.000000   NaN   
50%       NaN          NaN             2017.0      NaN    75.000000   NaN   
75%       NaN          NaN             2017.0      NaN   120.000000   NaN   
max       NaN          NaN             2017.0      NaN  9011.000000   NaN   

          odometer_km  registration_month fuel_type       brand  \
count     1452.000000         1452.000000       927        1452   
unique            NaN                 NaN         5          37   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN       652         378   
mean    132396.694215            4.621212       NaN         NaN   
std      35546.918201            3.892567       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            0.000000       NaN         NaN   
50%     150000.000000            4.000000       NaN         NaN   
75%     150000.000000            8.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count                823                 1452          1452.0   1452.000000   
unique                 2                   39             NaN           NaN   
top                 nein  2016-03-28 00:00:00             NaN           NaN   
freq                 689                   66             NaN           NaN   
mean                 NaN                  NaN             0.0  48597.771350   
std                  NaN                  NaN             0.0  25472.612226   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  28806.750000   
50%                  NaN                  NaN             0.0  47097.000000   
75%                  NaN                  NaN             0.0  66782.000000   
max                  NaN                  NaN             0.0  99947.000000   

                  last_seen  
count                  1452  
unique                 1447  
top     2016-04-06 12:46:49  
freq                      2  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [30]:
reg_year_2018.describe(include="all")

date_crawled             name  seller offer_type         price  \
count                   491              491     491        491    491.000000   
unique                  491              482       1          1           NaN   
top     2016-03-14 19:54:54  Mitsubishi_colt  privat    Angebot           NaN   
freq                      1                2     491        491           NaN   
mean                    NaN              NaN     NaN        NaN   3801.967413   
std                     NaN              NaN     NaN        NaN   5732.594754   
min                     NaN              NaN     NaN        NaN      0.000000   
25%                     NaN              NaN     NaN        NaN   1225.000000   
50%                     NaN              NaN     NaN        NaN   2100.000000   
75%                     NaN              NaN     NaN        NaN   4225.000000   
max                     NaN              NaN     NaN        NaN  64900.000000   

         abtest vehicle_type  registration_year  gearbox      power_ps model  \
count       491            1              491.0      432    491.000000   416   
unique        2            1                NaN        2           NaN   102   
top     control        coupe                NaN  manuell           NaN  golf   
freq        256            1                NaN      365           NaN    42   
mean        NaN          NaN             2018.0      NaN    113.953157   NaN   
std         NaN          NaN                0.0      NaN    721.974970   NaN   
min         NaN          NaN             2018.0      NaN      0.000000   NaN   
25%         NaN          NaN             2018.0      NaN      0.000000   NaN   
50%         NaN          NaN             2018.0      NaN     75.000000   NaN   
75%         NaN          NaN             2018.0      NaN    118.000000   NaN   
max         NaN          NaN             2018.0      NaN  16011.000000   NaN   

          odometer_km  registration_month fuel_type       brand  \
count      491.000000          491.000000       281         491   
unique            NaN                 NaN         5          34   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN       199         116   
mean    130000.000000            4.841141       NaN         NaN   
std      36869.875984            3.955778       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            1.000000       NaN         NaN   
50%     150000.000000            4.000000       NaN         NaN   
75%     150000.000000            8.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count                300                  491           491.0    491.000000   
unique                 2                   35             NaN           NaN   
top                 nein  2016-03-14 00:00:00             NaN           NaN   
freq                 280                   28             NaN           NaN   
mean                 NaN                  NaN             0.0  45579.867617   
std                  NaN                  NaN             0.0  24703.929419   
min                  NaN                  NaN             0.0   1454.000000   
25%                  NaN                  NaN             0.0  26650.000000   
50%                  NaN                  NaN             0.0  44147.000000   
75%                  NaN                  NaN             0.0  64293.000000   
max                  NaN                  NaN             0.0  99974.000000   

                  last_seen  
count                   491  
unique                  491  
top     2016-03-30 21:44:58  
freq                      1  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN

With the exception of 1 row for each year these listings seem to contain more than 1 incorrect value, so I've decided to remove these listings from the dataset.

In [31]:
autos = autos[autos["registration_year"].between(1910, 2016)]

In [32]:
# now let's look at the distribution of values in the registration_year column
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000687
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
          ...   
1987    0.001562
1988    0.002957
1989    0.003770
1990    0.008227
1991    0.007414
1992    0.008122
1993    0.009268
1994    0.013746
1995    0.027325
1996    0.030074
1997    0.042237
1998    0.051088
1999    0.062439
2000    0.069853
2001    0.056274
2002    0.052754
2003    0.056795
2004    0.057003
2005    0.062793
2006    0.056378
2007    0.047985
2008    0.046465
2009    0.043674
2010    0.033260
2011    0.034031
2012    0.027554
2013    0.016786
2014    0.0138

As you would expect there is a gradual increase in the number of listings for each year up to the year 2000 after which you can see a gradual decrease up to 2016. 2016 on the other hand shows a dramatic increase in the number of listings.

# Brands

All right, so at this point the guide for this project says to start aggregating the dataset by brands in order to analyse it. Which doesn't really make sense, since we haven't done everything we listed to do at the start. There are still a few columns we can drop and I highly doubt there is a car listed with 17700 horsepower. I could still tke those steps of course, but this project is quite boring and I really want to move on with the course. So here we go, analyzing the set by brand.

In [33]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [34]:
autos["brand"].value_counts().sort_index()

alfa_romeo          318
audi               4149
bmw                5283
chevrolet           274
chrysler            176
citroen             668
dacia               123
daewoo               72
daihatsu            123
fiat               1242
ford               3350
honda               377
hyundai             473
jaguar               76
jeep                108
kia                 341
lada                 29
lancia               52
land_rover           98
mazda               727
mercedes_benz      4579
mini                415
mitsubishi          391
nissan              725
opel               5194
peugeot            1418
porsche             292
renault            2274
rover                65
saab                 77
seat                873
skoda               770
smart               668
sonstige_autos      523
subaru              105
suzuki              284
toyota              599
trabant              75
volkswagen        10185
volvo               444
Name: brand, dtype: int64

In [35]:
brands = autos["brand"].value_counts().sort_index().index
brands

Index(['alfa_romeo', 'audi', 'bmw', 'chevrolet', 'chrysler', 'citroen',
       'dacia', 'daewoo', 'daihatsu', 'fiat', 'ford', 'honda', 'hyundai',
       'jaguar', 'jeep', 'kia', 'lada', 'lancia', 'land_rover', 'mazda',
       'mercedes_benz', 'mini', 'mitsubishi', 'nissan', 'opel', 'peugeot',
       'porsche', 'renault', 'rover', 'saab', 'seat', 'skoda', 'smart',
       'sonstige_autos', 'subaru', 'suzuki', 'toyota', 'trabant', 'volkswagen',
       'volvo'],
      dtype='object')

As we can see all values in the brand column are brands with the exception of sonstige_autos, which translates to 'other cars'. I've decided to look only at the brands and calculate the mean price for each brand. I've sorted the values and created an object to hold the brand names to aggregate by, but first I have to remove the sonstige_autos label.

In [36]:
brands = brands.drop("sonstige_autos")

In [39]:
# lets aggregate

mean_price_brand = {}

for b in brands:
    b_rows = autos[autos["brand"] == b]
    b_mean = int(b_rows["price"].mean())
    mean_price_brand[b] = b_mean

mean_price_brand

{'alfa_romeo': 3984,
 'audi': 9093,
 'bmw': 8102,
 'chevrolet': 6488,
 'chrysler': 3229,
 'citroen': 3699,
 'dacia': 5915,
 'daewoo': 1019,
 'daihatsu': 1556,
 'fiat': 2711,
 'ford': 3652,
 'honda': 3988,
 'hyundai': 5308,
 'jaguar': 11176,
 'jeep': 11434,
 'kia': 5789,
 'lada': 2502,
 'lancia': 3246,
 'land_rover': 19108,
 'mazda': 4010,
 'mercedes_benz': 8485,
 'mini': 10460,
 'mitsubishi': 3333,
 'nissan': 4664,
 'opel': 2876,
 'peugeot': 3039,
 'porsche': 43507,
 'renault': 2395,
 'rover': 1528,
 'saab': 3211,
 'seat': 4296,
 'skoda': 6334,
 'smart': 3542,
 'subaru': 3765,
 'suzuki': 3995,
 'toyota': 5115,
 'trabant': 1552,
 'volkswagen': 5231,
 'volvo': 4757}

A quick analysis of these numbers tells us that asian cars are cheapest, followed by European cars. German cars seem quite decently priced, with Porsche as the exception of course. Landrover also sticks out with a mean price over double that of nearly all other brands. It must be noted that I did not adjust anything to take into account the difference in the number of listings per brand.

## Let's make it a little more interesting before we move on

I'm going to first select the top 5 brands, then calculate the mean mileage and mean price and store the values in seperate dictionairies. From there I'll convert both to series objects and then to a combined dataframe object so I can analyse them next to each other.

In [42]:
top_5_brands  = (autos["brand"]
                 .value_counts()
                 .head(5)
                 .index
                )

top_5_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi'], dtype='object')

In [43]:
mean_price_top_5 = {}

for b in top_5_brands:
    b_rows = autos[autos["brand"] == b]
    b_mean = int(b_rows["price"].mean())
    mean_price_top_5[b] = b_mean

mean_mileage_top_5 = {}

for b in top_5_brands:
    b_rows = autos[autos["brand"] == b]
    b_mean = int(b_rows["odometer_km"].mean())
    mean_mileage_top_5[b] = b_mean

mean_price_top_5

{'audi': 9093,
 'bmw': 8102,
 'mercedes_benz': 8485,
 'opel': 2876,
 'volkswagen': 5231}

In [44]:
mean_mileage_top_5

{'audi': 129287,
 'bmw': 132431,
 'mercedes_benz': 130856,
 'opel': 129223,
 'volkswagen': 128724}

In [45]:
# coverting to series objects
mpt_5_series = pd.Series(mean_price_top_5)
mpt_5_series

audi             9093
bmw              8102
mercedes_benz    8485
opel             2876
volkswagen       5231
dtype: int64

In [46]:
mmt_5_series = pd.Series(mean_mileage_top_5)
mmt_5_series

audi             129287
bmw              132431
mercedes_benz    130856
opel             129223
volkswagen       128724
dtype: int64

In [47]:
top_5_mean_df = pd.DataFrame(mpt_5_series, columns=["mean_price"])
top_5_mean_df

mean_price
audi                 9093
bmw                  8102
mercedes_benz        8485
opel                 2876
volkswagen           5231

In [48]:
top_5_mean_df["mean_mileage"] = mmt_5_series

In [49]:
print(top_5_mean_df)

               mean_price  mean_mileage
audi                 9093        129287
bmw                  8102        132431
mercedes_benz        8485        130856
opel                 2876        129223
volkswagen           5231        128724


There seems to be no correlation between the mean mileage and mean price of a specific brand of car on this site. The mean mileage is around 130,000 km (yes, expressing mileage in kilometers is strangely ironic) for all brands, eventhough the mean price strongly differs.